### PR TITLE
Use SNMP GET for direct metrics

### LINKED ISSUE #131: Allowing SNMP get as an alternative to walk
**Use case:** Walking even a very short list of OID subtrees takes a very long time (~45 seconds for 10 subtrees) because of the number of leaf nodes in the tree. (Also it does not do pretty things to the device's CPU.) We only care about a very specific subset of OIDs for each target (the most important interfaces per switch), so taking 3/4 of a minute to fetch a lot of metrics before unceremoniously dropping most of them in relabeling is not ideal.

**Current solution:** One module per target, which lists _exactly_ the OIDs we want for that target.

If we want to get `ifHCInOctets` and `ifHCOutOctets` for ifIndex 555 and 666 on switch_one, then the config looks like:

```
switch_one:
  walk:
  - <ifHCInOctets OID>.555
  - <ifHCOutOctets OID>.555
  - <ifHCInOctets OID>.666
  - <ifHCOutOctets OID>.666
<label config is copied verbatim for each switch module>
```

This is not very scalable. (If there's a better way of achieving this with the exporter as-is, please point me to it. https://github.com/prometheus/snmp_exporter/issues/128 tells me this isn't currently possible.)

**Proposed functionality**: A `get` option in the configuration, that could work like so:

```
default:
  get:
  - <ifHCInOctets OID>.*
  - <ifHCOutOctets OID>.*
  - <someFullOIDWithoutWildcardIfDesired>
```

(I believe the config generator would have to generate a `get`-type module differently; rather than listing a minimal number of subtrees; it would list full OIDs but replace the final index with a wildcard.)

If a list of desired indexes were passed as a parameter along with the target, the exporter would replace the * with each index and then exclusively get the desired metrics per target without unnecessarily walking the full subtree to get a couple of metrics. (and without huge amounts of duplication in the config file)

I'd be happy to try to implement this (and likely will try to implement it on my end regardless), but would like to know if this would be a desired addition to the core exporter, and if you have any design suggestions that would help keep this in line with the design philosophy of Prometheus and the existing exporter.

### PR DESCRIPTION
**Pull request by sgissi**
Allows snmp_exporter to scrape targets using SNMP GET instead of GETBULK walks for direct metrics or specific table indexes. Fixes #131

### CODE REVIEW

File: generator/tree.go [hunk 1]
```diff
@@ -150,26 +150,103 @@ func minimizeOids(oids []string) []string {
...
 	out := &config.Module{}
 	needToWalk := map[string]struct{}{}
+	tableInstances := map[string][]string{}
 
 	// Remove redundant OIDs to be walked.
 	toWalk := []string{}
 	for _, oid := range cfg.Walk {
-		node, ok := nameToNode[oid]
-		if !ok {
+		oidNumber := getOid(oid, node, nameToNode)
```
[the remark this answers is not in the document]
Comment (sgissi) [reply generator/tree.go]: It does indeed, but this is the step where names are translated to numeric OIDs to remove overlaps/duplicates. Handling the duplicates after will require tracking back the removed entries to clean-up the tableInstances map. I think it is cleaner to way it is.

File: generator/FORMAT.md [hunk 2]
```diff
@@ -8,8 +8,11 @@ module_name:
   # There's various auth/version options here too. See the main README.
   walk:
     # List of OID subtrees to walk.
-    - 1.3.6.1.2.1.1.3
     - 1.3.6.1.2.1.2
+    - 1.3.6.1.2.1.31.1.1
+  get:
+    # List of OIDs to get directly.
+    - 1.3.6.1.2.1.1.3
```
[the remark this answers is not in the document]
Comment (sgissi) [reply generator/FORMAT.md]: It doesn't actually. The examples are in the subtree of 1.3.6.1.2.1, walking 2 (interfaces), 31.1.1 (ifXtable) and getting 1.3 (sysUptime). I can come up with other examples in other trees if it is less confusing.

**Comment by brian-brazil** [comment]
The final thing you'll need to do is update the existing snmp.yml with these changes, as I presume there's diffs.

File: snmp.yml [hunk 3]
```diff
@@ -1953,6 +1955,7 @@ cisco_wlc:
   - 1.3.6.1.4.1.14179.2.2.2.1.2
   - 1.3.6.1.4.1.14179.2.2.2.1.4
   - 1.3.6.1.4.1.14179.2.2.6.1
+  get: []
```
Comment (brian-brazil) [inline snmp.yml]: Hmm, are we missing an omitempty?

Comment (sgissi) [reply snmp.yml]: Yes indeed, added for both Get and Walk, and cleaned up snmp.yml.

**Comment by brian-brazil** [comment]
Thanks!

**Comment by brian-brazil** [comment]
I was just testing this out and if I get an oid that the device doesn't support I get 0 rather than no sample. This should be fixed.

**Comment by sgissi** [comment]
Yes, the BulkWalk will remove invalid responses but Get won't, I opened #287 to filter that. Also found an issue that scalar OIDs (like sysUpTime, sysLocation, etc.) must be always accessed via index 0, SNMP has weird rules sometimes.

**Comment by RichiH** [comment]
Given that we usually omit trailing `.0`, it should probably be added by generator. The only question is if generator should fail if `.0` is added in its config.

**Comment by brian-brazil** [comment]
We already don't work if the generator config has .0 appended. Noone has complained.

**Pull request merged**